### PR TITLE
Loading spinner when running editor queries and fetching the schema

### PR DESCRIPTION
**Pull request by asiandrummer**
A CSS-driven loading spinner when running editor queries and fetching the schema, in case they take a while. Shown in a result window and documentation explorer (since it needs a schema to be rendered).

### CODE REVIEW

File: src/components/GraphiQL.js [hunk 1]
```diff
@@ -199,6 +200,10 @@ export class GraphiQL extends React.Component {
 
   componentDidMount() {
     if (!this.state.schema) {
+      this.setState({
+        showSpinner: true
```
Comment (leebyron) [inline src/components/GraphiQL.js]: We should only show a spinner when there is an user-produced query to the server underway, not for the automated fetching of the schema. GraphiQL is still usable without a schema, just typeahead and linters are not yet enabled, that's fine.

This implementation also has a race condition where we could issue a query before the schema returns. The schema would return and turn off the spinner, but the query itself hasn't returned yet.

File: src/components/GraphiQL.js [hunk 2]
```diff
@@ -171,6 +171,7 @@ export class GraphiQL extends React.Component {
       variableEditorHeight: this._storageGet('variableEditorHeight') || 200,
       docsOpen: false,
       docsWidth: this._storageGet('docExplorerWidth') || 350,
+      showSpinner: false,
```
Comment (leebyron) [inline src/components/GraphiQL.js]: I'd recommend calling this `isWaitingForResponse` since thats the semantic value, and the spinner is just the visualization of this value.

File: src/components/GraphiQL.js [hunk 3]
```diff
@@ -297,6 +308,13 @@ export class GraphiQL extends React.Component {
               </div>
             </div>
             <div className="resultWrap">
+              {(() => {
+                return this.state.showSpinner && (
+                  <div className="spinner-container">
+                    <div className="spinner" />
+                  </div>
+                );
+              })()}
```
Comment (leebyron) [inline src/components/GraphiQL.js]: why the IIFE? couldn't you also write:

```
{this.state.showSpinner &&
  <div className="spinner-container">
    <div className="spinner" />
  </div>}
```

File: src/components/GraphiQL.js [hunk 4]
```diff
@@ -297,6 +308,13 @@ export class GraphiQL extends React.Component {
...
               <ResultViewer ref="result" value={this.state.response} />
```
Comment (leebyron) [inline src/components/GraphiQL.js]: Perhaps the ResultViewer should be grayed out or something while another query is underway and the spinner is placed over it.

File: src/components/ResultViewer.js [hunk 5]
```diff
@@ -67,6 +67,8 @@ export class ResultViewer extends React.Component {
   }
 
   render() {
-    return <div className="result-window" />;
+    return (
```
Comment (leebyron) [inline src/components/ResultViewer.js]: unnecessary change

**Comment by leebyron** [comment]
You should also squash your commits together. You can do that with `rebase -i`, and then `push -f` to update the PR

**Comment by asiandrummer** [comment]
screen shot 2015-11-25 at 3 01 34 pm

File: css/loading.css [hunk 6]
```diff
@@ -0,0 +1,62 @@
+#graphiql-container .spinner-container {
+  position: absolute;
+  top: 20%;
+  height: 36px;
+  width: 36px;
+  left: 50%;
+  transform: translate(-50%,-50%);
```
Comment (leebyron) [inline css/loading.css]: nit: space after the comma

File: css/loading.css [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  border-left: 6px solid rgba(150, 150, 150, .15);
+  border-right: 6px solid rgba(150, 150, 150, .15);
+  border-bottom: 6px solid rgba(150, 150, 150, .15);
+  border-top: 6px solid rgba(150, 150, 150, .8);
+  border-radius: 100%;
+}
+
+@-webkit-keyframes rotation {
+  from {
+  -webkit-transform: rotate(0deg);
```
Comment (leebyron) [inline css/loading.css]: nit: indent inside the {} or one-line it. I think it may be nicer to read:

```css
from { -crap-transform: rotate(0deg); }
to { -crap-transform: rotate(359deg); }
```

File: src/components/GraphiQL.js [hunk 8]
```diff
@@ -329,11 +338,18 @@ export class GraphiQL extends React.Component {
...
+      this.setState({
+        isWaitingForResponse: false,
+        response: error && error.stack || error
+      });
     });
   }
 
   _runEditorQuery = () => {
+    this.setState({
+      isWaitingForResponse: true,
```
Comment (leebyron) [inline src/components/GraphiQL.js]: this should also set `response: null`

File: css/app.css [hunk 9]
```diff
@@ -115,6 +116,10 @@ html, body {
   border-left: solid 1px #e0e0e0;
 }
 
+#graphiql-container .resultWrap.isWaitingForResponse .CodeMirror {
+  background: rgb(200, 200, 200);
```
Comment (leebyron) [inline css/app.css]: I don't think you need this, the spinner looks nicer on a white background IMHO

File: css/loading.css [hunk 10]
```diff
@@ -0,0 +1,62 @@
+#graphiql-container .spinner-container {
+  position: absolute;
+  top: 20%;
```
Comment (leebyron) [inline css/loading.css]: 50% would be nicer to vertically center
